trieste/space.py

```diff
@@ -331,3 +331,149 @@ def __mul__(self, other: Box) -> Box:
 
     def __deepcopy__(self, memo: dict[int, object]) -> Box:
         return self
+
+
+class OrdinalSearchSpace(Box):
+    r"""
+    Ordinal search space representing a discrete :math:`D`-dimensional box in
+    :math:`\mathbb{R}^D`. This class inherit :class:`Box` and generates evenly-spaced values
+    in an interval for each dimension. For example randomly sample from this search space
+    will give:
+
+        >>> tf.random.set_seed(1945)
+        >>> ordinalsp = OrdinalSearchSpace([-2.0, -2.0], [2.0, 2.0], [0.2, 0.3])
+        >>> sample = ordinalsp.sample(5)
+        >>> sample.numpy()
+        array([[ 0. , -0.6],
+               [ 0.6, -0.3],
+               [-1.2,  1.2],
+               [-0.4,  1.5],
+               [ 1.2,  0. ]])
+
+    """
+
+    def __init__(
+        self,
+        lower: Sequence[float] | TensorType,
+        upper: Sequence[float] | TensorType,
+        stepsizes: Sequence[float] | TensorType,
+    ):
+
+        r"""
+        If ``lower``, ``upper`` and ``stepsizes`` are `Sequence`\ s of floats
+        (such as lists or tuples), they will be converted to tensors of dtype `tf.float64`.
+        if ``lower`` and ``upper`` are not multiple of ``stepsizes`` it will be replaced by
+        smallest and largest multiples inside previous ``lower`` and ``upper``.
+
+        :param lower: The lower (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param upper: The upper (inclusive) bounds of the Ordinal Space. Must have shape [D]
+            for positive D, and if a tensor, must have float type.
+        :param stepsizes: The step size of each dimension of the Ordinal Space. Must have
+            shape [D] for positive D, and if a tensor, must have float type.
+        :raise ValueError (or InvalidArgumentError): If any of the following are true:
+
+            - ``lower``, ``upper`` and ``stepsizes`` have invalid shapes.
+            - ``lower``, ``upper`` and ``stepsizes`` do not have the same floating point type.
+            - ``upper`` is not greater than ``lower`` across all dimensions.
+        """
+
+        super().__init__(lower, upper)
+        tf.debugging.assert_shapes([(lower, ["D"]), (stepsizes, ["D"])])
+        tf.debugging.assert_positive(stepsizes)
+        tf.assert_rank(stepsizes, 1)
+
+        tf.debugging.assert_positive(len(stepsizes), message="stepsizes cannot be empty")
+
+        if isinstance(stepsizes, Sequence):
+            self._stepsizes = tf.constant(stepsizes, dtype=tf.float64)
+        else:
+            self._stepsizes = tf.convert_to_tensor(stepsizes)
+
+            tf.debugging.assert_same_float_dtype([self._lower, self._stepsizes])
+
+        self._lower = tf.math.ceil(tf.divide(self._lower, self._stepsizes)) * self._stepsizes
```

Comment:
FYI I think this correctly handles floating point errors: `tf.math.ceil` rounds 5.0000001 to 5, not 6, as it's very close to 5. Therefore, while `math.ceil((5/3)/(1/3))` return `6`, `tf.math.ceil(tf.divide(5/3, 1/3))` returns 5.

Maybe update `test_ordinalsearchspace_bounds_changes_to_nearest_multiples` so it checks `OrdinalSearchSpace([5/3],[2.0],[1/3])` too.

Reply:
checked on python yes it does work, but strangely that test fail (maybe because of dtype?). 
fixed by subtracting the term inside ceil with small jitter though:
 `self._lower = tf.math.ceil(tf.divide(self._lower, self._stepsizes) - 1e-6) * self._stepsizes`